server/src/main/java/org/jboss/as/server/deployment/DeploymentUnitPhaseService.java

```diff
@@ -208,19 +171,34 @@ public void transition(final ServiceController<?> controller, final ServiceContr
             phaseServiceBuilder.install();
         }
     }
+    public synchronized void stop(final StopContext context) {
+        ServiceName top = deploymentUnit.getParent() != null ? deploymentUnit.getParent().getServiceName() : deploymentUnit.getServiceName();
+        ServiceController<?> topController = context.getController().getServiceContainer().getService(top);
+        if(topController.getMode() != Mode.REMOVE && topController.getMode() != Mode.NEVER && !context.getController().getServiceContainer().isShutdown()) {
+            ServerLogger.DEPLOYMENT_LOGGER.deploymentRestartDetected(deploymentUnit.getName());
+            //the deployment is going down, but it has not been explicitly stopped or removed
+            //so it must be because of a missing dependency. Unfortunately these phase services cannot fully restart
+            //as the data they require no longer exists, so instead we trigger a complete redeployment
+            //the redeployment will likely not fully complete, but will be waiting on whatever missing dependency
+            //caused this stop
 
-    private Boolean restartAllowed() {
-        final DeploymentUnit parent;
-        if (deploymentUnit.getParent() == null) {
-            parent = deploymentUnit;
-        } else {
-            parent = deploymentUnit.getParent();
-        }
-        Boolean allowed = parent.getAttachment(Attachments.ALLOW_PHASE_RESTART);
-        return allowed != null && allowed;
-    }
+            //we get the deployment unit service name
+            //add a listener to perform a restart when the service goes down
+            //then stop the deployment unit service
+            topController.addListener(new AbstractServiceListener<Object>() {
 
-    public synchronized void stop(final StopContext context) {
+                @Override
+                public void transition(final ServiceController<?> controller, final ServiceController.Transition transition) {
+                    if(transition.getAfter().equals(ServiceController.Substate.DOWN)) {
+                        //its possible an undeploy happened in the meantime
+                        //so we use compareAndSetMode to make sure the mode is still NEVER and not REMOVE
+                        controller.compareAndSetMode(Mode.NEVER, Mode.ACTIVE);
+                        controller.removeListener(this);
+                    }
+                }
+            });
+            topController.setMode(Mode.NEVER);
```

Comment:
I think this needs to be a CAS as well or again it can race with an undeploy op. It perhaps could be moved to the if test.